APL record support

fcelda opened the pull request:
This PR adds support for [APL record type](https://tools.ietf.org/html/rfc3123).

miekg commented on types.go:
```diff
@@ -1402,11 +1483,24 @@ func euiToString(eui uint64, bits int) (hex string) {
 }
 
+// copyBytes returns a copy of a byte slice.
+func copyBytes(b []byte) []byte {
```
this is a too thin of a wrapper of `copy` right now. Please just keep copyIP and copy the netmask below manually

tmthrgd commented on duplicate_generate.go:
```diff
@@ -100,6 +100,16 @@ func main() {
 
+				if st.Tag(i) == `dns:"apl"` {
+					o3(`for i := 0; i < len(r1.%s); i++ {
+						if r1.%s[i].Equals(&r2.%s[i]) {
```
This should be not equals (`!...`).

fcelda replied:
Good catch!

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
 	return off, nil
 }
+
+var aplAddrLenToFamily = map[int]uint16{
```
Just use a switch statement for these. I'm pretty sure we don't use maps anywhere else.

fcelda replied:
This is used in four different places so there will be some duplication but I can do that.

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+
+func packDataApl(data []APLPrefix, msg []byte, off int) (int, error) {
+	var err error
+	for i, _ := range data {
```
Drop the `_`.

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+		return len(msg), err
+	}
+	// Write AFDPART
+	copy(msg[off:], addr)
```
`off += copy(...)`.

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+			return nil, len(msg), err
+		}
+		off = end
+		result = append(result, *prefix)
```
I don't like that `unpackDataAplPrefix` returns a pointer but `unpackDataApl` returns a slice of values. Pick one and stick to it throughout.

fcelda replied:
This was a practical decision. Most of the parsing happens in `unpackDataAplPrefix`, there are eight return statements (currently after addressing other review comments), and returning nil pointer is more convenient than returning an empty structure. I would actually prefer leaving it this way. What do you think? Should I keep it or change it?

tmthrgd replied:
`APLPrefix{}` isn’t much longer than `nil` and this inconsistency isn’t great.

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+
+	// Read ADDRESSFAMILY
+	var family uint16
+	family, off, _ = unpackUint16(msg, off)
```
These can all use the short variable declaration (`:=`) instead of having a separate `var`.

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+
+	// Read ADDRESSFAMILY
+	var family uint16
+	family, off, _ = unpackUint16(msg, off)
```
I don't like that we're ignoring the errors from `unpackUintX` in this function.

fcelda replied:
The only reason for the unpack functions to fail is buffer underflow and I was checking the header boundaries a few lines above. But I'm going to change this for consistency with the existing code. :+1:

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+	// Read N and AFDLENGTH
+	var nlen uint8
+	nlen, off, _ = unpackUint8(msg, off)
+	neg := (nlen & 0x80) > 0
```
`!= 0`

tmthrgd commented on msg_helpers.go:
```diff
@@ -681,3 +681,135 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+		return nil, len(msg), &Error{err: "overflow unpacking APL address"}
+	}
+	ip := make([]byte, ipLen)
+	copy(ip, msg[off:off+afdlen])
```
`off += copy(...)`

tmthrgd commented on msg_helpers_test.go:
```diff
@@ -153,3 +157,309 @@ func BenchmarkUnpackString(b *testing.B) {
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ap := APLPrefix{
```
Put the `&` here instead of at `packDataAplPrefix`.

tmthrgd commented on msg_helpers_test.go:
```diff
@@ -153,3 +157,309 @@ func BenchmarkUnpackString(b *testing.B) {
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ap := APLPrefix{Network: net.IPNet{IP: tt.ip, Mask: tt.mask}}
```
`&` here as well.

tmthrgd commented on msg_helpers_test.go:
```diff
@@ -153,3 +157,309 @@ func BenchmarkUnpackString(b *testing.B) {
+}
+
+func TestPackDataAplPrefix_BufferBounds(t *testing.T) {
+	ap := APLPrefix{
```
`&` here as well.

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+
+// String returns presentation form of the APL record.
+func (rr *APL) String() string {
+	sb := strings.Builder{}
```
`var sb strings.Builder`

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+	sb := strings.Builder{}
+	sb.WriteString(rr.Hdr.String())
+	for i, p := range rr.Prefixes {
+		if i != 0 {
```
`i > 0`

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+
+// String returns presentation form of the APL prefix.
+func (p *APLPrefix) String() string {
+	sb := strings.Builder{}
```
`var sb strings.Builder`

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+		if i != 0 {
+			sb.WriteByte(' ')
+		}
+		sb.WriteString(p.String())
```
Having a `String() string` method on both `APL` and `APLPrefix` both of which allocate isn't exactly great. Do we really need one on `APLPrefix`?

fcelda replied:
Maybe not. I will make it private method. :+1:

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+	case net.IPv6len:
+		// add prefix for IPv4-mapped IPv6
+		if v4 := p.Network.IP.To4(); v4 != nil {
+			sb.WriteString("::ffff:")
```
Does `net.IP.String()` not do this already?

fcelda replied:
Unfortunately no. IPv4-mapped IPv6 is as IPv4 per [IP.String](https://godoc.org/net#IP.String).

I'm not sure what are the reasons but Go often stores IPv4 as IPv6 address and then performs conversions in the background so nobody would notice. The conversion to string is one example of that.

APL record is protocol-agnostic and we need to respect the address type and therefore I added the prefix to force the presentation format to be IPv6. I just checked the specification again whether the IPv4-mapped IPv6 is allowed and it is. The other implementation option I considered was to print the address as a pure IPv6 address but that would require different kind of trickery. I would stick to this solution.

tmthrgd replied:
Ah I see. Don’t love it, but I guess it’s okay. Is there anything wrong with just converting IPv4 as IPv6 address to IPv4 throughout?

fcelda replied:
I think there are the following options to write IPv4 APL in presentation format:

1. `1:192.0.2.0/24`
2. `2:192.0.2.0/24` (invalid)
3. `2:::ffff:192.0.2.0/120`
4. `2:::ffff:c000:0200/120`

Option 1 is plain IPv4. Option 2 is invalid according to the specification, we could be lax and allow it in this library but that would make it incompatible with the other libraries. Options 3 and 4 are equivalent on the wire.

Even though all the valid options are semantically the same (i.e. refer to the same network subnet), the APL has explicit notation of address family so there is an actual difference between 1 and 3 (or 4) in the wire format.

I think we cannot force conversion from IPv4-mapped IPv6 to IPv4 because then part of the information is lost. It would also lead to confusing behavior when converting from write to presentation and back to wire again.

Therefore we should stick to option 3 or 4. Go doesn't have API to print address in that format so we can either write a custom conversion function or use the trick above. For option 4, I also have this if it's prefered:
```go
ip := net.ParseIP("::ffff:192.0.2.1")
fmt.PrintlN(ip) // 192.0.2.1
ip[1] = 1 // hack
fmt.Println(ip.String()[1:]) // ::ffff:c000:201
```

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+			sb.WriteString("::ffff:")
+		}
+		sb.WriteString(p.Network.IP.String())
+	default:
```
Drop the `default` label.

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+// Equals reports whether two APL prefixes are identical.
+func (a *APLPrefix) Equals(b *APLPrefix) bool {
+	return a.Negation == b.Negation &&
+		bytes.Equal(a.Network.IP, b.Network.IP) &&
```
Use `net.IP.Equal` here.

fcelda replied:
I cannot because IPv4 and equivalent IPv4-mapped IPv6 would be treated as the same (see [IP.Equal](https://godoc.org/net#IP.Equal)) while the APL resource record would be different.

tmthrgd commented on types.go:
```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+}
+
+// Len returns size of the prefix in wire format.
+func (p *APLPrefix) Len() int {
```
How useful is this method? If it's not really needed externally it should be either unexported on preferably merged into  types_generate.go.

fcelda replied:
I'm going to make all methods on APLPrefix private. I can imagine some of them may become handy but probably not much. :+1:

tmthrgd commented on types_generate.go:
```diff
@@ -249,6 +251,12 @@ func main() {
 					continue
 				}
+				if t == "APLPrefix" {
+					fmt.Fprintf(b, "%s := make([]%s, len(rr.%s));\nfor i, _ := range rr.%s {\n %s[i] = rr.%s[i].Copy()\n}\n",
```
Drop the `_` here.

miekg commented:
APL RFC https://tools.ietf.org/html/rfc3123 should also be added to the README

fcelda commented:
@miekg @tmthrgd Thank you for looking at my changes thoroughly. I've addressed all the comments you left in code or made my case in the comments. It should be ready for another round of reviews.

miekg commented on parse_test.go:
```diff
@@ -1584,3 +1584,137 @@ func TestNULLRecord(t *testing.T) {
+	}
+}
+
+func TestParseAPL_Errors(t *testing.T) {
```
No underscores please.

miekg commented on types.go:
```diff
@@ -1353,6 +1355,91 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+	case net.IPv6len:
+		afi = 2
+	}
+	sb.WriteString(strconv.Itoa(int(afi)))
```
this can be written inside the switch, no need for `afi` integer then

miekg commented on types.go:
```diff
@@ -1353,6 +1355,91 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
+
+// len returns size of the prefix in wire format.
+func (p *APLPrefix) len() int {
+	// 4-byte header and the network address prefix (see https://tools.ietf.org/html/rfc3123#section-4)
```
reference the RFC by the number , not a link to tools.ietf.org

fcelda commented:
OK, should be fixed now.

miekg commented on msg_helpers.go:
```diff
@@ -681,3 +681,129 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+	addr := p.Network.IP.Mask(p.Network.Mask)[:(prefix+7)/8]
+
+	// Write ADDRESSFAMILY
+	off, err := packUint16(family, msg, off)
```
sorry to keep on nittig: but this can also put in the switch, no need for `family` var, just try to do the write there. The comment also doesn't *really* add useful context here.

miekg commented on msg_helpers.go:
```diff
@@ -681,3 +681,129 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+	if err != nil {
+		return len(msg), err
+	}
+	// Write PREFIX
```
comment doesn't add anything

miekg commented on msg_helpers.go:
```diff
@@ -681,3 +681,129 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+	// Write N and AFDLENGTH
+	var n uint8
+	if p.Negation {
+		n = 0x80
```
these constants come from the RFC?

fcelda replied:
Yes, these are the names from the [RFC](https://tools.ietf.org/html/rfc3123#section-4). I can remove that. It's not super helpful.

```
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
      |                          ADDRESSFAMILY                        |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
      |             PREFIX            | N |         AFDLENGTH         |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
      /                            AFDPART                            /
      |                                                               |
      +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
```

miekg commented on msg_helpers.go:
```diff
@@ -681,3 +681,129 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
+		return APLPrefix{}, len(msg), &Error{err: "unrecognized APL address family"}
+	}
+	// Read PREFIX
+	prefix, off, err := unpackUint8(msg, off)
```
read prefix before the `switch family` and then do the `int(prefix)> *ipLen` check there, this way to you drop the `ipLen` variable.

I see `ipLen` is also used as the very last thing in this function, probably with building up a APLPrefix as you go, to not have to variables, but just initialize an APLPrefix along the way.

fcelda replied:
I'm not sure I understood exactly what you proposed. I moved all the reads of the fixed-size header to the beginning of the function so that the address length handling logic is together. `ipLen` is gone and so is `neg`. Please, take a look.

fcelda commented:
OK, folks, another round. :upside_down_face:

miekg commented:
It's not super important and we can always fix it later. So

/lgtm

On Thu, 19 Dec 2019, 19:44 Jan Včelák, <notifications@github.com> wrote:

> OK, folks, another round. 🙃
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/miekg/dns/pull/1058?email_source=notifications&email_token=AACWIW6TPUZ7PWWU7JARQMLQZPFJNA5CNFSM4J4MHOP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHKWYTQ#issuecomment-567635022>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACWIW7ZTNIH6JYBH2SMWPLQZPFJNANCNFSM4J4MHOPQ>
> .
>

miekg commented:
I'm ready to merge this cc @tmthrgd

fcelda commented:
Thank you for the the review and patience, @miekg and @tmthrgd.
